src/core/shape_inference/include/shape_infer_type_utils.hpp

```diff
@@ -11,18 +11,26 @@ namespace ov {
 namespace util {
 
 /**
- * \brief Trnsform tensor data by cast them to type T
+ * \brief Transform tensor data by cast them to type T
  *
  * \tparam T Type of returned value.
  */
 template <class T>
 struct Cast {
     constexpr Cast() = default;
 
-    template <class U>
+    template <
+        class U,
+        typename std::enable_if<!std::is_integral<T>::value || !std::is_floating_point<U>::value>::type* = nullptr>
     constexpr T operator()(const U u) const {
         return static_cast<T>(u);
     }
+
+    template <class U,
+              typename std::enable_if<std::is_integral<T>::value && std::is_floating_point<U>::value>::type* = nullptr>
+    constexpr T operator()(const U u) const {
+        return u < std::numeric_limits<T>::max() ? static_cast<T>(u) : std::numeric_limits<T>::max();
```

Comment:
I think we also can get the same issue in case of lower values... Am I wrong?

Reply:
Yes you've got right it can be same issue for negative values.
Added additional check.